enhancements/run-bundle.md

```diff
@@ -0,0 +1,263 @@
+
+The idea is to consolidate the run command into different subcommands to make
+things easier to understand and eliminate flag bloat.
+
+ * `run local` - exists today
+ * `run packagemanifest` - covered by [OSDK-1126](https://issues.redhat.com/browse/OSDK-1126)
+ * `run bundle` - covered by this design document
+
+#### Command Line Interface
+The new command line will look as follows:
+
+```
+operator-sdk run bundle <bundle-image> [--index-image=] [--namespace=] [--install-mode=(AllNamespace|OwnNamespace|SingleNamespace=)]
+```
+
+* `<bundle-image>` is a positional argument that specifies the bundle image.
+* `[--index-image=]` is an optional flag that specifies the index image.
+  * If `--index-image` is not provided, default to
+    `quay.io/operator-framework/upstream-registry-builder`
+  * If `--index-image` is provided, use the provided value
+* `[--namespace=]` is an optional flag that specifies the namespace to install
+  the operator. It will default to the `KUBECONFIG` context.
+* `[--install-mode=]` is an optional flag that specifies the
+  [OLM install mode](https://red.ht/2YK0ksD).
+  * We support 3 modes:
+    * AllNamespaces
+    * OwnNamspace
+    * SingleNamespace=ns1
+      * NOTE: if ns1 is the same as the operator's namespace, a warning will be
+        printed
+  * When `--install-mode` is not supplied, the default precendence order is:
+    * `OperatorGroup`, if it already exists
+    * `AllNamespaces`, if the CSV supports it
+    * `OwnNamespace`, if the CSV supports it
+    * If the CSV does not support either of these, `--install-mode` flag
+      will be **required**
+
+#### Assumptions
+Like the existing `run` command, we **assume** the following:
+
+ 1. OLM is already installed
+ 1. kube API is available with the permissions to create resources
+ 1. kube server can see and pull the provided images (i.e. bundle and index)
+ 1. If an index image is provided, we assume that it has:
+    * bash, so we can change the entrypoint
+    * dependencies required for `opm registry add` so we can inject
+       the bundle into the DB
+    * the existing registry DB located at `/database/index.db`
+
+#### Example UX Scenarios
+
+Running the following scenarios results in different results. While not an
+exhaustive list, these are some example uses of the `run bundle` command.
+
+* `operator-sdk run bundle <bundle-image>` would install into the default kube namespace,
+creates a cluster-wide `OperatorGroup`.
+* `operator-sdk run bundle <bundle-image> --namespace foo` would install into
+  namespace `foo`, creates a cluster-wide `OperatorGroup`
+* `operator-sdka run bundle <bundle-image> --namespace foo --install-mode
+  SingleNamespace=bar` would install into namespace foo, would determine if
+  OperatorGroup exists and is sufficient, if not it would error
+* `operator-sdk run bundle <bundle-image> --install-mode SingleNamespace=bar`
+  would install into the default kube namespace, does the OperatorGroup logic
```

Comment:
I know you've prefaced that this list is not exhaustive, but since we're planning to have logic that looks at what the operator supports and whether there's an existing OperatorGroup, would it make sense to describe the setup and then say what happens when different invocations are run?

Something like:

> - Given an operator that supports install modes `OwnNamespace` and `SingleNamespace` being installed into a namespace with no existing operator group:
>
>     - `operator-sdk run bundle <bundle-image>` would install into the namespace configured in the `KUBECONFIG` context and an OperatorGroup would be created, targeting the same install namespace
>     - `operator-sdk run bundle <bundle-image> --namespace foo` would install into the `foo` namespace and an OperatorGroup would be created, targeting the `foo` namespace
>    - `operator-sdk run bundle <bundle-image> --namespace foo --install-mode
  SingleNamespace=bar` would install into namespace `foo`, and an OperatorGroup would be created, targeting the `bar` namespace
>
> - Given an operator that supports AllNamespaces, OwnNamespace, and Single namespace being installed into a namespace with an existing OperatorGroup that targets all namespaces:
>
>     - `operator-sdk run bundle <bundle-image>` would install into the namespace configured in the `KUBECONFIG` context and the operator would watch all namespaces.
>     - `operator-sdk run bundle <bundle-image> --namespace foo` would install into the `foo` namespace and the operator would watch all namespaces.
>     - `operator-sdk run bundle <bundle-image> --namespace foo --install-mode
  SingleNamespace=bar` would install into namespace `foo`, and the operator would watch the `bar` namespace.

Also, not sure if that last bullet in the second scenario is actually correct. If the `OperatorGroup` targets all namespaces, is it possible for an operator installed in that namespace to be configured to watch just one namespace?

Reply:
I redid the entire Example UX Scenarios section. I don't want to list out every possible combination. I'm not entirely sure which ones are valid. Here's what I ended up with.
#### Example UX Scenarios

Running the following scenarios results in different results. While not an
exhaustive list, these are some example uses of the `run bundle` command.

* `operator-sdk run bundle <bundle-image>`
* `operator-sdk run bundle <bundle-image> --namespace foo`
* `operator-sdk run bundle <bundle-image> --namespace foo --install-mode SingleNamespace=bar`
* `operator-sdk run bundle <bundle-image> --install-mode SingleNamespace=bar`

It is important to remember that the presence of an OperatorGroup in the 
namespace can affect the how `run bundle` behaves. Let's look at a couple scenarios.

* Given an operator that supports install modes `OwnNamespace` and 
  `SingleNamespace` being installed into a namespace with *no* OperatorGroup.
  The `run bundle` command would behave as follows:
  * `operator-sdk run bundle <bundle-image>` would install into the default
    namespace configured in the `KUBECONFIG` context. An OperatorGroup would be
    created targeting the same install namespace.
  * `operator-sdk run bundle <bundle-image> --namespace foo --install-mode SingleNamespace=bar`
    would install into namespace `foo`. An OperatorGroup would be created
    targeting the `bar` namespace.
* Given an operator that supports install modes `AllNamespaces`, `OwnNamespace`,
  and `SingleNamespace`, being installed into a namespace with an existing
  OperatorGroup that targets all namespaces:
  * `operator-sdk run bundle <bundle-image>` would install into the default
    namespace configured in the `KUBECONFIG` context. The operator would watch
    all namespaces.
  * `operator-sdk run bundle <bundle-image> --namespace foo --install-mode SingleNamespace=bar`
    would install into namespace `foo`. The operator would watch namespace
    `bar`.

As you can see the presence of the OperatorGroup will take precedence. If the 
OperatorGroup does not support a version an error condition will occur. The `run
bundle` command will return with a non-zero exit status and log the errors.